btk/metrics.py

```diff
@@ -1,484 +1,497 @@
-    num_shred1 = len(np.where(true_table["num_detections1"] > 1)[0])
-    num_detected2 = len(np.where(true_table["num_detections2"] == 1)[0])
-    num_undetected2 = len(np.where(true_table["num_detections2"] == 0)[0])
-    num_spurious2 = len(np.where(det_table["match_true_id2"] == -1)[0])
-    num_shred2 = len(np.where(true_table["num_detections2"] > 1)[0])
-    if not num_detected1 + num_undetected1 + num_shred1 == num_true:
-        raise ValueError(
-            "Number of detected objects + number undetected "
-            "objects must be equal to the total number of true "
-            "objects"
-        )
+    match_table = astropy.table.Table()
+    t_x = true_table["x_peak"].reshape(-1, 1) - detected_table["x_peak"].reshape(1, -1)
+    t_y = true_table["y_peak"].reshape(-1, 1) - detected_table["y_peak"].reshape(1, -1)
+    dist = np.hypot(t_x, t_y)  # dist[i][j] = distance between true object i and detected object j.
 
-    if not num_detected2 + num_undetected2 + num_shred2 == num_true:
-        raise ValueError(
-            "Number of detected objects + number undetected "
-            "objects must be equal to the total number "
-            "of true objects."
-        )
+    # solve optimization problem.
+    # true_table[true_indx[i]] is matched with detected_table[detected_indx[i]]
+    # len(true_indx) = len(detect_indx) = min(len(true_table), len(detected_table))
+    true_indx, detected_indx = linear_sum_assignment(dist)
 
-    num_matched_detections1 = true_table["num_detections1"].sum()
-    if not num_matched_detections1 + num_spurious1 == num_det:
-        raise ValueError(
-            "Number of detections match to a true object + "
-            "number of spurious must be equal to the "
-            "total number of detections."
-        )
+    # for each true galaxy i, match_indx[i] is the index of detected_table matched to that true
+    # galaxy or -1 if there is no match.
+    match_indx = [-1] * len(true_table)
+    dist_m = [0.0] * len(true_table)
+    for i, indx in enumerate(true_indx):
+        match_indx[indx] = detected_indx[i]
+        dist_m[indx] = dist[indx][detected_indx[i]]
 
-    num_matched_detections2 = true_table["num_detections2"].sum()
-    if not num_matched_detections2 + num_spurious2 == num_det:
-        raise ValueError(
-            "Number of detections match to a true object + "
-            "number of spurious must be equal to the total "
-            "number of detections."
-        )
+    match_table["match_detected_id"] = match_indx
+    match_table["dist"] = dist_m
+    return match_table
 
-    blend_summary = [
-        num_true,
-        num_detected1,
-        num_undetected1,
-        num_spurious1,
-        num_shred1,
-        num_detected2,
-        num_undetected2,
-        num_spurious2,
-        num_shred2,
-    ]
-    return blend_summary
 
+def detection_metrics(blend_list, detection_catalogs, matches):
```

Comment:
Thanks for making this function more "global" @thuiop - so now it's calculated using all the information from a single batch. I think we might want to make it even more "global" and calculate it only over several batches at once (so in `run_metrics`, again because we can't simply average all the f1-scores over all batches right?). 

I think we can do it by including a `detection_summary` table which for now we can just have columns `blend_id` and `matched`.

Reply:
I guess this depends on the `batch_size`, I imagine we probably want to keep the batch_size small, and use multiprocessing ?

Reply:
actually if we return the `true_pos`, `false_pos`, etc. for each blend/batch that should be enough?

Reply:
If we return the raw numbers we can compute the metrics on several batches yes. Although I do not see why you would particularly want that compared to a single batch, except maybe if you want to compute the data in several times.

Reply:
I feel like the most common workflow is to have `batch_size` relatively small so that when doing multiprocessing each `cpu` doesn't have a lot of work to do in the `draw_blend_generator` and in the `measure_generator`. At the end multiple batches can be combined to get a large dataset.  Am I misunderstanding something? This is also why I implemented the `run_metrics` function.

Do let me know if you think an alternative workflow would be better though @thuiop

Reply:
The batch size is not the size of the batches for multiprocessing ; when you set `cpus` to a value larger than 1, the batch is divided in several minibatches of size batch_size/cpus and then recombined as a single batch later. You may then want to generate several batches using the next method repeatedly if you want several batches.

Reply:
That's true. Anyways I think this is not too important, I'll let you handle how you think the best way for users to interact with the `metrics_generator` is. Feel free to remove the `run_metrics` function if you don't find it useful.